subscriber.go

```diff
@@ -61,12 +80,23 @@ func newBlockingSubscriber(cap int) *blockingSubscriber {
 	}
 
 	return &blockingSubscriber{
-		ch: make(chan Message, cap),
+		ch:      make(chan Message, cap),
+		closeCh: make(chan struct{}),
 	}
 }
 
 // Set will send the message using the channel.
 func (s *blockingSubscriber) Set(msg Message) {
+	s.chMu.RLock()
+	defer s.chMu.RUnlock()
+
+	// check s.ch isn't closed (we are holding the RLock, so s.ch won't be closed until the end of this function)
+	select {
```

Comment:
You are using the closed boolean on the nonBlockingSubscriber, It's not possible to use the same here?
I think it's cleaner and makes them consistent.

Reply:
I read your comment and found a bug in my implementation as well :)
Consider the following sequence of events:

1. Set takes RLock
2. s.closed is checked inside Set (s.closed == false)
3. Set blocks at select for a long time
4. Close is called, it sets s.closed to true
5. Close tries to take Lock, but it can't, because Set is still running
6. Deadlock

In other words, to take the lock, Close has to stop all blocking operations that have RWLock, but afaik the only way to stop a blocking operation is with a select

Reply:
About the problem of closing a channel written by several goroutines I think we have two solutions:
First, the solution described on https://www.leolara.me/blog/closing_a_go_channel_written_by_several_goroutines/: https://gist.github.com/leolara/f6fb5dfc04d64947487f16764d6b37b6
The only thing I would change is the send method not use a goroutine because this method is already been called by a goroutine and I don't want to make them async and without control.

The other solution is to change the hub and make the Unsubscribe and Publish concurrent safe using an read lock for publish (read the subscribers) and the Close, Subscribe, Unsubscribe use write lock. 

WDYT?